File: app/src/main/java/com/example/kocja/rabbiter_online/activities/viewEntry.java
package com.example.kocja.rabbiter_online.activities;

import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.example.kocja.rabbiter_online.managers.GsonManager;
import com.example.kocja.rabbiter_online.managers.HttpManager;
import com.example.kocja.rabbiter_online.R;
import com.example.kocja.rabbiter_online.databases.Entry;
import com.example.kocja.rabbiter_online.databases.Events;
import com.example.kocja.rabbiter_online.fragments.HistoryFragment;
import com.example.kocja.rabbiter_online.fragments.viewEntryData;
import com.google.gson.Gson;

import java.util.UUID;


/**
 * Created by kocja on 27/01/2018.
 */

public class viewEntry extends AppCompatActivity {
    private Entry mainEntry;
    private viewEntryData mainEntryFragment;
    private boolean dataChanged = false;
    private ImageView mainEntryView;
    private Gson gson;

    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view_entry);
        Intent currentIntent = getIntent();

        gson = GsonManager.getGson();
        UUID mainEntryUUID = (UUID) currentIntent.getSerializableExtra("entryID");
        mainEntryView = findViewById(R.id.mainEntryView);
        Intent viewLargerImage = new Intent(this,largerMainImage.class);
        mainEntryView.setOnClickListener(view -> {
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
                mainEntryView.setTransitionName("closerLook");
                ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(this,mainEntryView,"closerLook");
                startActivity(viewLargerImage,options.toBundle());
            }
            else{
                startActivity(viewLargerImage);
            }

        });
        HttpManager.postRequest("seekSingleEntry",gson.toJson(mainEntryUUID), (response, bytes) -> {

            this.runOnUiThread(() -> {
                Entry entry = gson.fromJson(response,Entry[].class)[0];
                viewLargerImage.putExtra("imageURI", entry.getEntryPhLoc());
                mainEntry = entry;
                mainEntryFragment = (viewEntryData) getSupportFragmentManager().findFragmentById(R.id.mainEntryFragment);

                RecyclerView historyView = findViewById(R.id.upcomingAdapter);

                if(entry.getChooseGender().equals(getString(R.string.genderMale))){
                    HistoryFragment.maleParentOf(this, entry.getEntryName(),historyView,viewEntry.this);
                }
                else {
                    HistoryFragment.setPastEvents(this,entry.getEntryName(),historyView);
                }

                mainEntryFragment.setData(entry);

                HttpManager.postRequest("searchForImage", gson.toJson(entry.getEntryPhLoc()), (response1, bytes1) -> {
                    entry.setEntryBitmap(BitmapFactory.decodeByteArray(bytes1,0, bytes1.length));
                    this.runOnUiThread(() -> Glide.with(viewEntry.this).load(entry.getEntryBitmap()).into(mainEntryView));

                });


                if(entry.isMerged()){
                    ImageView mergedView = findViewById(R.id.mergedView);
                    mergedView.setOnClickListener(view -> {
                        Intent startMergedMain = new Intent(this,viewEntry.class);
                        startMergedMain.putExtra("entryID",entry.getMergedEntry());
                        ActivityOptionsCompat compat = ActivityOptionsCompat.makeSceneTransitionAnimation(viewEntry.this,mainEntryView,"mergedName");
                        startActivity(startMergedMain,compat.toBundle());
                    });
                    mergedView.setVisibility(View.VISIBLE);
                    HttpManager.postRequest("searchForImage", gson.toJson(entry.getMergedEntryPhLoc()), (response1, bytes1) -> {
                        entry.setMergedEntryBitmap(BitmapFactory.decodeByteArray(bytes1,0, bytes1.length));
                        this.runOnUiThread(()->Glide.with(viewEntry.this).load(entry.getMergedEntryBitmap()).into(mergedView));
                    });

                }
            });

        });
    }
    public void onActivityResult(int requestCode,int resultCode,Intent data){
        if(requestCode == addEntryActivity.EDIT_EXISTING_ENTRY && resultCode == RESULT_OK){
            HttpManager.postRequest("seekSingleEntry", gson.toJson(mainEntry), (response,bytes) -> {
                Entry entry = gson.fromJson(response,Entry[].class)[0];
                this.runOnUiThread(() -> {
                    mainEntryFragment.setData(entry);
                    Glide.with(viewEntry.this)
                            .load(entry.getEntryPhLoc())
                            .into(mainEntryView);
                });

            });
            dataChanged = true;
        }
    }
    public void onBackPressed(){
        if(dataChanged){
            setResult(RESULT_OK);
            supportFinishAfterTransition();
            //finish();
        }
        else{
            super.onBackPressed();
        }
    }
    public boolean onCreateOptionsMenu(Menu menu){
        getMenuInflater().inflate(R.menu.menu_view_entry_data,menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.editEntry) {
            Intent startEditProc = new Intent(viewEntry.this, addEntryActivity.class);
            startEditProc.putExtra("getMode", addEntryActivity.EDIT_EXISTING_ENTRY);
            startEditProc.putExtra("entryEdit", mainEntry.getEntryID());
            startActivityForResult(startEditProc, addEntryActivity.EDIT_EXISTING_ENTRY);

        } else if (id == R.id.deleteEntry) {
            AlertDialog.Builder assureDeletion = new AlertDialog.Builder(viewEntry.this)
                    .setTitle(R.string.confirmDeletion)
                    .setPositiveButton(R.string.confirm, (dialogInterface, i) -> HttpManager.postRequest("seekEventsName", gson.toJson(mainEntry.getEntryName()), (response, bytes) -> {
                        Events[] events = gson.fromJson(response,Events[].class);
                        this.runOnUiThread(() -> {
                            for (Events event : events) {
                                HttpManager.postRequest("deleteEvent", gson.toJson(event.getEventUUID()), (response1,bytes1) -> {
                                });
                            }
                            HttpManager.postRequest("deleteEntry", gson.toJson(mainEntry.getEntryID()), (response1,bytes1) -> {
                                setResult(RESULT_OK);
                                finish();
                            });
                        });


                    }))
                    .setNegativeButton(R.string.decline, (dialogInterface, i) -> dialogInterface.cancel());
            assureDeletion.show();

        }
        else if (id == R.id.entryStats) {
            Intent startStatActivity = new Intent(getApplicationContext(), viewEntryStats.class);
            startStatActivity.putExtra("entryUUID", mainEntry.getEntryID());
            startActivity(startStatActivity);

        }
        else if(id == R.id.showMerged){
            if(mainEntry.isMerged()){
                Intent showMerged = new Intent(getApplicationContext(),viewEntry.class);
                showMerged.putExtra("entryID",UUID.fromString(mainEntry.getMergedEntry()));
                startActivity(showMerged);
            }
            else{
                AlertDialog.Builder alertNotMerged = new AlertDialog.Builder(this)
                        .setTitle("Oops")
                        .setMessage("your entry is not merged")
                        .setPositiveButton("OK", (dialogInterface, i) -> dialogInterface.cancel());
                alertNotMerged.show();
            }

        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/kocja/rabbiter_online/activities/ViewEntry.kt
package com.example.kocja.rabbiter_online.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle

import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityOptionsCompat
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.lifecycle.lifecycleScope
import coil.load

import com.example.kocja.rabbiter_online.R
import com.example.kocja.rabbiter_online.databinding.ActivityViewEntryBinding
import com.example.kocja.rabbiter_online.fragments.HistoryFragment
import com.example.kocja.rabbiter_online.models.Entry
import com.example.kocja.rabbiter_online.viewmodels.ViewEntryViewModel
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel


/**
 * Created by kocja on 27/01/2018.
 */

class ViewEntry : AppCompatActivity() {
    private var dataChanged = false
    private val viewEntryViewModel: ViewEntryViewModel by viewModel()
    private lateinit var binding : ActivityViewEntryBinding
    private val viewLargerImage by lazy{ Intent(this, LargerMainImage::class.java)}

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityViewEntryBinding.inflate(layoutInflater)
        setContentView(binding.root)
        val mergedEntryUUID = intent.getStringExtra("mergedEntryUUID")

        binding.mainEntryView.setOnClickListener {
            binding.mainEntryView.transitionName = "closerLook"
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, binding.mainEntryView, "closerLook")
            startActivity(viewLargerImage, options.toBundle())
        }
        if(mergedEntryUUID != null){
            lifecycleScope.launch{
                val result = viewEntryViewModel.findEntryByUUID(mergedEntryUUID)
                initViewEntry(result)
            }
        }
        else initViewEntry(intent.getParcelableExtra("entry")!!)
    }

    private fun initViewEntry(entry : Entry){
        viewEntryViewModel.entry.value = entry
        viewLargerImage.putExtra("imageURL", entry.entryPhotoURL)

        val historyFragment = supportFragmentManager.findFragmentById(R.id.historyFragment) as HistoryFragment

        if (entry.chooseGender == getString(R.string.genderMale)) historyFragment.maleParentOf(this, entry.entryName, historyFragment.fragmentUpcomingHistoryLayoutBinding!!.upcomingAdapter)
        else historyFragment.setPastEvents(this, entry.entryName, historyFragment.fragmentUpcomingHistoryLayoutBinding!!.upcomingAdapter)

        binding.mainEntryView.load(viewEntryViewModel.entry.value!!.entryPhotoURL)


        if (entry.isMerged) {
            binding.mergedView.setOnClickListener {
                val startMergedMain = Intent(this, ViewEntry::class.java)
                startMergedMain.putExtra("entryUUID", entry.mergedEntryID)
                val compat = ActivityOptionsCompat.makeSceneTransitionAnimation(this@ViewEntry, binding.mainEntryView, "mergedName")
                startActivity(startMergedMain, compat.toBundle())
            }
            binding.mergedView.visibility = View.VISIBLE
            binding.mergedView.load(viewEntryViewModel.entry.value?.mergedEntryPhotoURL)
        }
    }

    override fun onBackPressed() {
        if (dataChanged) {
            setResult(Activity.RESULT_OK,Intent().putExtra("updatedEntry",viewEntryViewModel.entry.value!!))
            supportFinishAfterTransition()
        } else super.onBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_view_entry_data, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.editEntry -> {
                val result = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
                    if (it.resultCode == Activity.RESULT_OK) {
                        viewEntryViewModel.entry.value = it.data?.getParcelableExtra("updatedEntry")

                        binding.mainEntryView.load(viewEntryViewModel.entry.value!!.entryPhotoURL)
                        dataChanged = true
                    }
                }
                val addEntry = Intent(this@ViewEntry, AddEntryActivity::class.java)
                addEntry.putExtra("getMode", AddEntryActivity.EDIT_EXISTING_ENTRY)
                addEntry.putExtra("entry", viewEntryViewModel.entry.value!!)
                result.launch(addEntry)
            }
            R.id.deleteEntry -> {
                val entryUUID = viewEntryViewModel.entry.value!!.entryUUID
                val areYouSure = AlertDialog.Builder(this@ViewEntry)
                        .setTitle(R.string.confirmDeletion)
                        .setPositiveButton(R.string.confirm) { _, _ ->
                            lifecycleScope.launch {
                                val result = viewEntryViewModel.findEventsName()
                                result.forEach{
                                    viewEntryViewModel.deleteEvent(it.eventUUID)
                                }
                                viewEntryViewModel.deleteEntry()
                                setResult(Activity.RESULT_OK, Intent().putExtra("deletedEntryUUID", entryUUID))
                                finish()
                            }
                        }
                        .setNegativeButton(R.string.decline) { dialogInterface, _ -> dialogInterface.cancel() }
                areYouSure.show()
            }
            R.id.entryStats -> {
                val startStatActivity = Intent(applicationContext, ViewEntryStats::class.java)
                startStatActivity.putExtra("entry", viewEntryViewModel.entry.value)
                startActivity(startStatActivity)
            }
            R.id.showMerged -> {
                if (viewEntryViewModel.entry.value!!.isMerged) {
                    val showMerged = Intent(applicationContext, ViewEntry::class.java)
                    showMerged.putExtra("mergedEntryUUID", viewEntryViewModel.entry.value?.mergedEntryID)
                    startActivity(showMerged)
                } else {
                    val alertNotMerged = AlertDialog.Builder(this)
                            .setTitle("Oops")
                            .setMessage("your entry is not merged")
                            .setPositiveButton("OK") { dialogInterface, _ -> dialogInterface.cancel() }
                    alertNotMerged.show()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
